config-model-api/src/main/java/com/yahoo/config/model/api/Quota.java

```diff
@@ -14,42 +16,57 @@
  * @author ogronnesby
  */
 public class Quota {
+    private static final Quota UNLIMITED = new Quota(Optional.empty(), Optional.empty());
 
+    /** The max size of a cluster inside application */
     private final Optional<Integer> maxClusterSize;
 
     /** The max budget in dollars per hour */
-    private final Optional<Integer> budget;
+    private final Optional<BigDecimal> budget;
 
-    public Quota(Optional<Integer> maybeClusterSize, Optional<Integer> budget) {
-        this.maxClusterSize = maybeClusterSize;
-        this.budget = budget;
+    // TODO: Remove constructor once Vespa < 7.300 is gone from production
+    public Quota(Optional<Integer> maxClusterSize, Optional<Integer> budget) {
+        this(maxClusterSize, budget.map(BigDecimal::new), true);
+    }
+
+    private Quota(Optional<Integer> maxClusterSize, Optional<BigDecimal> budget, boolean isDecimal) {
+        this.maxClusterSize = Objects.requireNonNull(maxClusterSize);
+        this.budget = Objects.requireNonNull(budget);
     }
 
     public static Quota fromSlime(Inspector inspector) {
         var clusterSize = SlimeUtils.optionalLong(inspector.field("clusterSize"));
-        var budget = SlimeUtils.optionalLong(inspector.field("budget"));
-        return new Quota(clusterSize.map(Long::intValue), budget.map(Long::intValue));
+        var budget = budgetFromSlime(inspector.field("budget"));
+        return new Quota(clusterSize.map(Long::intValue), budget, true);
+    }
+
+    public Quota withBudget(Optional<BigDecimal> budget) {
```

Comment:
Consider `BigDecimal` parameter instead of `Optional`. That way you avoid the initial problem where signatures with generic types cannot be overloaded. Same for the other method.